lang/cs/Org.Apache.REEF.Network/Elastic/Task/NodeObserverIdentifier.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Network.Elastic.Comm.Impl;
+using Org.Apache.REEF.Network.Elastic.Topology.Physical;
+using Org.Apache.REEF.Utilities.Attributes;
+
+namespace Org.Apache.REEF.Network.Elastic.Task.Impl
+{
+    /// <summary>
+    /// An identifier for a given node in the group communication topology.
+    /// A node is uniquely identifiable by a combination of its 
+    /// <see cref="StageName"/>, and <see cref="OperatorId"/>.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    internal sealed class NodeObserverIdentifier
+    {
+         /// <summary>
+        /// Creates an identifier from an operator topology with communication.
+        /// </summary>
+        public static NodeObserverIdentifier FromObserver(IOperatorTopologyWithCommunication observer)
+        {
+            return new NodeObserverIdentifier(observer.StageName, observer.OperatorId);
+        }
+
+        /// <summary>
+        /// Creates an from a driver aware topology.
+        /// </summary>
+        public static NodeObserverIdentifier FromObserver(DriverAwareOperatorTopology observer)
+        {
+            return new NodeObserverIdentifier(observer.StageName, observer.OperatorId);
+        }
+
+        /// <summary>
+        /// Creates an identifier from a group communication message.
+        /// </summary>
+        public static NodeObserverIdentifier FromMessage(ElasticGroupCommunicationMessage message)
+        {
+            return new NodeObserverIdentifier(message.StageName, message.OperatorId);
+        }
+
+        /// <summary>
+        /// Basic constructor.
+        /// </summary>
+        /// <param name="stageName">The name of the stage</param>
+        /// <param name="operatorId">The identifier of the operator</param>
+        private NodeObserverIdentifier(string stageName, int operatorId)
+        {
+            StageName = stageName;
+            OperatorId = operatorId;
+        }
```

Comment:
I think we can just have a struct and have all comparisons, hash, etc. for free. we can also enforce the interface that provides stage name + operator id for all classes that we use as `FromMessage()` parameters here. Say, we define
```C#
public interface INodeIdentifier
{
    /// <summary>The stage name.</summary>
    string StageName { get; }

    /// <summary>The operator name.</summary>
    int OperatorId { get; }
}
```
that we implement in `ElasticGroupCommunicationMessage` and other classes, and also here in our struct:
```C#
public struct NodeIdentifier : INodeIdentifier
{
    /// <summary>The stage name.</summary>
    public string StageName { get; private set; }

    /// <summary>The operator name.</summary>
    public int OperatorId { get; private set; }

    public NodeIdentifier(string stageName, int operatorId)
    {
        StageName = stageName;
        OperatorId = operatorId;
    }
}
```
if we want to pack that tuple, e.g., as a key, `ElasticGroupCommunicationMessage` can instead implement something like
```C#
public interface IIdentifiable : INodeIdentifier
{
    NodeIdentifier NodeId { get; }
}
```
either way, we won't need to define our own `GetHashCode()`, `Equals()`, and custom `FromMessage()` conversions.

Reply:
I think I got you mean. This is indeed cleaner. Fixed.